src/trio/testing/_raises_group.py

```diff
@@ -0,0 +1,341 @@
+    def __init__(self, *, check: Callable[[BaseException], bool]):
+        ...
+
+    def __init__(
+        self,
+        exception_type: type[E] | None = None,
+        match: str | Pattern[str] | None = None,
+        check: Callable[[E], bool] | None = None,
+    ):
+        if exception_type is None and match is None and check is None:
+            raise ValueError("You must specify at least one parameter to match on.")
+        if exception_type is not None and not issubclass(exception_type, BaseException):
+            raise ValueError(
+                f"exception_type {exception_type} must be a subclass of BaseException"
+            )
+        self.exception_type = exception_type
+        self.match: Pattern[str] | None
+        if isinstance(match, str):
+            self.match = re.compile(match)
+        else:
+            self.match = match
+        self.check = check
+
+    def matches(self, exception: BaseException) -> TypeGuard[E]:
+        if self.exception_type is not None and not isinstance(
+            exception, self.exception_type
+        ):
+            return False
+        if self.match is not None and not re.search(
+            self.match, _stringify_exception(exception)
+        ):
+            return False
+        # If exception_type is None check() accepts BaseException.
+        # If non-none, we have done an isinstance check above.
+        if self.check is not None and not self.check(cast(E, exception)):
+            return False
+        return True
+
+    def __str__(self) -> str:
+        reqs = []
+        if self.exception_type is not None:
+            reqs.append(self.exception_type.__name__)
+        if (match := self.match) is not None:
+            # If no flags were specified, discard the redundant re.compile() here.
+            reqs.append(
+                f"match={match.pattern if match.flags == _regex_no_flags else match!r}"
+            )
+        if self.check is not None:
+            reqs.append(f"check={self.check!r}")
+        return f'Matcher({", ".join(reqs)})'
+
+
+# typing this has been somewhat of a nightmare, with the primary difficulty making
+# the return type of __enter__ correct. Ideally it would function like this
+# with RaisesGroup(RaisesGroup(ValueError)) as excinfo:
+#   ...
+# assert_type(excinfo.value, ExceptionGroup[ExceptionGroup[ValueError]])
+# in addition to all the simple cases, but getting all the way to the above seems maybe
+# impossible. The type being RaisesGroup[RaisesGroup[ValueError]] is probably also fine,
+# as long as I add fake properties corresponding to the properties of exceptiongroup. But
+# I had trouble with it handling recursive cases properly.
+
+# Current solution settles on the above giving BaseExceptionGroup[RaisesGroup[ValueError]], and it not
+# being a type error to do `with RaisesGroup(ValueError()): ...` - but that will error on runtime.
+if TYPE_CHECKING:
+    SuperClass = BaseExceptionGroup
+else:
+    SuperClass = Generic
```

Comment:
Why not just always inherit from `BaseExceptionGroup`?

And I feel like inheriting from that's super fishy given we don't actually have the information `BaseExceptionGroup` would have. But TBH I don't really see another solution. The typing on this is really icky.

Reply:
I gave typing a go and ran across:
```py
import contextlib
from typing import ContextManager, Generic, TypeVar, overload

E = TypeVar("E", bound=BaseException)

class FakeExcInfo(Generic[E]):
    value: E

class RaisesGroup(ContextManager[FakeExcInfo[BaseExceptionGroup[E]]], Generic[E]):
    @overload
    def __init__(self, a: type[E], *b: type[E]):
        ...
    
    @overload
    def __init__(self: RaisesGroup[BaseExceptionGroup[E]], a: RaisesGroup[E], *b: RaisesGroup[E]):
        ...
    
    def __init__(self: object, *args: object):
        ...
    
    def __exit__(self, *args: object) -> bool:
        ...


with RaisesGroup(RaisesGroup(ValueError)) as excinfo:
    pass

reveal_type(excinfo.value)
```

it has a couple issues (not accepting both `RaisesGroup` and normal errors, plus whatever else kind of exceptions we allow) and also mypy doesn't like the overloads idk why but like. maybe that's cleaner?

Reply:
For what it's worth, pyright seems to dislike that (probably a bug) but there's an easy workaround for it:
```py
from typing import ContextManager, Generic, TypeVar, overload

E = TypeVar("E", bound=BaseException)
E2 = TypeVar("E2", bound=BaseException)

class FakeExcInfo(Generic[E]):
    value: E

class RaisesGroup(ContextManager[FakeExcInfo[BaseExceptionGroup[E]]], Generic[E]):
    @overload
    def __init__(self, a: type[E], *b: type[E]):
        ...
    
    @overload
    def __init__(self: "RaisesGroup[BaseExceptionGroup[E2]]", a: "RaisesGroup[E2]", *b: "RaisesGroup[E2]"):
        ...
    
    def __init__(self: object, *args: object):
        ...
    
    def __exit__(self, *args: object) -> bool:
        ...


with RaisesGroup(RaisesGroup(ValueError)) as excinfo:
    pass

reveal_type(excinfo.value)
```

Reply:
> Why not just always inherit from `BaseExceptionGroup`?

This gives errors due to the funky `__new__` stuff that `BaseExceptionGroup` does.
```
>       with RaisesGroup(ValueError, KeyError):
E       TypeError: BaseExceptionGroup.__new__() argument 1 must be str, not type
```

implementing a `__new__` that works like `object.__new__`:
```
>       obj = object.__new__(cls)
E       TypeError: object.__new__(RaisesGroup) is not safe, use BaseExceptionGroup.__new__()
```

and since we don't actually care for any of the functionality in `BaseExceptionGroup`, not inheriting at runtime seems clearly correct.

Reply:
> And I feel like inheriting from that's super fishy given we don't actually have the information BaseExceptionGroup would have. But TBH I don't really see another solution. The typing on this is really icky.

Yeah this is clearly doing stuff that typing wasn't meant to handle.
The underlying problem is getting this to pass typing:
```python
with RaisesGroup(RaisesGroup(ValueError)) as excinfo:
  ...
myexc: BaseExceptionGroup = excinfo.value.exceptions[0]
```
and the only way I see to manage this is to lie to the type checker that it can treate `RaisesGroup` like a `BaseExceptionGroup`.

I'll update comments with reasons for some of the shady stuff.

Reply:
also see https://github.com/python-trio/trio/pull/2898#issuecomment-1848932771

Reply:
What did you think of my proposed workaround using overloads btw?

Rest makes sense though.

Reply:
Ugh I poked a bit more and the `E2` thing above that pyright accepts, mypy doesn't. I'm fine with this hacky inheritance thing I guess, hopefully inheritance doesn't fall under semver.

Reply:
I wouldn't think so for this particular case, because there would be no need to preform `isinstance` checks because you would not expect RaisesGroup to be in a list mixed with other objects.

Reply:
> What did you think of my proposed workaround using overloads btw?
> 
> Rest makes sense though.

Oh I didn't trying to run it and failed to grok at a glance how it resolved the issue, sorry! I'll try and understand it and see if I can get it to work, otherwise I'll stick with the ugly inheritance thing.

Reply:
The E2 version + pyright doesn't work for me with a third level of nesting
```
with RaisesGroup(RaisesGroup(RaisesGroup(ValueError))) as excinfo3:
    pass

reveal_type(excinfo3.value)
```
gives
```
information: Type of "excinfo3.value" is "BaseExceptionGroup[BaseExceptionGroup[Unknown]]"
```

Reply:
but yeah let's stick with the hacky inheritance for now, and we can revisit this if a version of this is accepted for upstream pyright.